### ISSUE TITLE
Right truncation can produce all-ignored labels and a NaN SFT loss

### ISSUE THREAD

**Issue by dajiaohuang** [bytedance/UniVR#13]
## Summary

`VRTrainDataset` truncates sequences from the right after constructing labels. If the masked prompt/problem prefix already fills `max_length`, every retained label is `-100`, so the sample has no supervised target and the standard cross-entropy reduction returns `NaN`.

Affected upstream commit: `3d1965cac5529b72c9dab9af7a3f456c5bbd7172`.

## Evidence / reproduction

In `UniVR_SFT/src/utils/dataset_vr_train.py` lines 689-707, the sequence is assembled as prompt/problem tokens followed by summary/answer tokens; the entire prefix is labeled `-100`; then both arrays are sliced as `[:self.max_length]`. There is no check that any target remains.

Minimal loss reproduction of the resulting label state:

```python
import torch
import torch.nn.functional as F

labels = torch.tensor([-100] * 4)  # result after right truncation
logits = torch.randn(4, 8)
print((labels != -100).sum().item())
print(F.cross_entropy(logits, labels, ignore_index=-100).item())
```

Observed with PyTorch 2.12.0.dev20260330 on Windows/Python 3.13:

```text
0
nan
```

This state is reachable when the user prompt plus problem-image tokens and assistant prefix are at least `max_length`; the answer-frame budget logic only reduces answer frames and forces at least one, but does not constrain the preceding problem-image/prompt portion.

## Expected behavior

Every emitted training sample should retain at least one supervised token, or the dataset should reject/skip it with a clear diagnostic before it reaches the loss.

## Impact

An oversized problem context can inject a non-finite loss into SFT, potentially destabilizing or aborting a distributed training run while giving no indication which sample caused it.

## Why there is no PR in this report

The safe policy is product/data dependent: crop problem frames, preserve the supervised tail, reject the sample, or change the maximum-length contract. Those choices change training semantics, and the official dataset/model-scale run cannot be validated in this Windows audit environment. A maintainer decision on the intended truncation policy is needed before adding the corresponding dataset regression test and implementation.

## Duplicate check

I reviewed all existing issues and pull requests and searched for label truncation, `ignore_index`, all-masked labels, and NaN loss reports; no equivalent report was present as of 2026-08-22.

**Comment by MaverickRen** [comment]
Thank you for the detailed report. We confirmed the issue in the current code: [train_sft_full.sh](https://github.com/bytedance/UniVR/blob/main/UniVR_SFT/scripts/train_sft_full.sh#L80-L84) passes --use_lora True, while train.py enables get_peft_model when use_lora is set.
Therefore, the full-parameter launcher currently performs LoRA fine-tuning instead of full-parameter fine-tuning. The fix is tracked in [PR #6](https://github.com/bytedance/UniVR/pull/6), which changes the full-parameter launcher to
 --use_lora False and adds a regression test to distinguish the two launchers. Thank you for identifying this issue.

**Comment by MaverickRen** [comment]
Thank you for the detailed report. We confirmed the failure mode in the current `VRTrainDataset` implementation. The sample is assembled with the BOS token, user prompt, problem-image tokens,
  assistant prefix, and BSS token first, and this entire prefix is assigned the ignore label `-100`. The supervised summary, answer-image tokens, ESS token, and EOS token are appended afterward
  in [`dataset_vr_train.py`](https://github.com/bytedance/UniVR/blob/main/UniVR_SFT/src/utils/dataset_vr_train.py#L689-L699). The original truncation logic then applies `input_ids[:max_length]`
  and `labels[:max_length]` independently at [`dataset_vr_train.py`](https://github.com/bytedance/UniVR/blob/main/UniVR_SFT/src/utils/dataset_vr_train.py#L701-L714). When the masked prompt and
  problem-image prefix already reaches `max_length`, this removes every supervised label and emits a sample whose labels are entirely `-100`. PyTorch’s standard cross-entropy reduction then has
  no valid target positions and returns `NaN`.

  We chose to fix the problem at the dataset boundary instead of modifying the loss function or silently replacing the non-finite loss. The normal right-truncation behavior is retained when the
  supervised region begins before `max_length`. If the masked prefix itself reaches or exceeds the limit, the dataset now keeps the supervised tail by removing the excessive left-side context.
  This guarantees that the emitted sample contains at least one non-ignored label, while still respecting the configured maximum sequence length. The behavior is implemented in a dedicated
  truncation helper so that the policy is explicit and independently testable.

  The helper also validates the result before the sample is returned. It rejects invalid non-positive sequence limits, verifies that `input_ids` and `labels` have matching lengths, and raises a
  clear error if truncation would still remove every supervised label rather than allowing an all-ignored sample to reach the trainer. Samples that require the tail-preserving path are marked
  with `prefix_removed=True` in `sample_info`, which makes the affected examples identifiable in training logs. This keeps the fix localized to sequence construction and does not change the
  model’s loss computation.

  A focused regression test covers the long-prefix case, verifies that the supervised tail is retained, confirms that ordinary truncation still preserves the existing prefix behavior when the
  target begins before the limit, and checks that an all-ignored result is rejected explicitly. The fix has been implemented in the local `fix/sft-truncation-target` branch and committed as
  `9da273b`. The relevant source behavior is documented in [`VRTrainDataset._build_training_sample`](https://github.com/bytedance/UniVR/blob/main/UniVR_SFT/src/utils/dataset_vr_train.py#L522-
  L775), and the issue is tracked in [Issue #13](https://github.com/bytedance/UniVR/issues/13).
